homeassistant/components/intellifire/config_flow.py

```diff
@@ -4,60 +4,117 @@
 import voluptuous as vol
 
 from homeassistant import config_entries
+from homeassistant.components.dhcp import DhcpServiceInfo
 from homeassistant.const import CONF_HOST
-from homeassistant.core import HomeAssistant
 from homeassistant.data_entry_flow import FlowResult
 
-from .const import DOMAIN
+from .const import DOMAIN, LOGGER
 
 STEP_USER_DATA_SCHEMA = vol.Schema({vol.Required(CONF_HOST): str})
 
 
-async def validate_input(hass: HomeAssistant, host: str) -> str:
+async def validate_host_input(host: str) -> str:
     """Validate the user input allows us to connect.
 
     Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
     """
     api = IntellifireAsync(host)
     await api.poll()
-
+    ret = api.data.serial
+    LOGGER.info("Found a fireplace: %s", ret)
     # Return the serial number which will be used to calculate a unique ID for the device/sensors
-    return api.data.serial
+    return ret
 
 
 class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
     """Handle a config flow for IntelliFire."""
 
-    VERSION = 1
+    VERSION = 3
 
-    async def async_step_user(
-        self, user_input: dict[str, Any] | None = None
-    ) -> FlowResult:
-        """Handle the initial step."""
-        if user_input is None:
-            return self.async_show_form(
-                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
-            )
-        errors = {}
+    def __init__(self):
+        """Initialize the Config Flow Handler."""
+        self._discovered_host: str = ""
+        self.fireplace_finder = AsyncUDPFireplaceFinder()
+        self._config_context = {}
 
+    async def async_step_dhcp(self, discovery_info: DhcpServiceInfo) -> FlowResult:
+        """Handle DHCP Discovery."""
+        potential_host = discovery_info.ip
+        LOGGER.debug(discovery_info)
         try:
-            serial = await validate_input(self.hass, user_input[CONF_HOST])
+            serial = await validate_host_input(potential_host)
         except (ConnectionError, ClientConnectionError):
-            errors["base"] = "cannot_connect"
-        else:
-            await self.async_set_unique_id(serial)
-            self._abort_if_unique_id_configured(
-                updates={CONF_HOST: user_input[CONF_HOST]}
-            )
-
-            return self.async_create_entry(
-                title="Fireplace",
-                data={CONF_HOST: user_input[CONF_HOST]},
-            )
+            return self.async_abort(reason="not_intellifire_device")
+
+        await self.async_set_unique_id(serial)
+        # check if found before
+        self._abort_if_unique_id_configured(
+            updates={
+                CONF_HOST: potential_host,
+            }
+        )
+
+        self._discovered_host = potential_host
+        return await self.async_step_user()
+
+    async def _find_fireplaces(self):
+        """Perform UDP discovery."""
+        ips = await self.fireplace_finder.search_fireplace(timeout=1)
```

Comment:
I think you can create it here instead

Reply:
Changes would be made in: https://github.com/home-assistant/core/pull/67002